src/core/command_markers.rs

```diff
@@ -0,0 +1,261 @@
+            commands,
+            entity_map,
+        };
+
+        let component: C = serde_fns.deserialize(cursor, &mut mapper)?;
+        let mut history = entity
+            .get_mut::<ComponentHistory<C>>()
+            .expect("entity should have the history since it's as a marker");
+
+        history.push(component);
+
+        Ok(())
+    }
+
+    /// Stores history of values of `C` received from server. Present only on client.
+    ///
+    /// In this example, we use it as both a marker and storage.
+    #[derive(Component, Deref, DerefMut)]
+    struct ComponentHistory<C>(Vec<C>);
+    ```
+    **/
+    unsafe fn register_marker_fns<M: Component, C: Component>(
+        &mut self,
+        write: WriteFn,
+        remove: RemoveFn,
+    ) -> &mut Self;
+}
+
+impl AppMarkerExt for App {
+    fn register_marker<M: Component>(&mut self) -> &mut Self {
+        self.register_marker_with_priority::<M>(0)
+    }
+
+    fn register_marker_with_priority<M: Component>(&mut self, priority: usize) -> &mut Self {
+        let component_id = self.world.init_component::<M>();
+        let mut command_markers = self.world.resource_mut::<CommandMarkers>();
+        let marker_id = command_markers.insert(CommandMarker {
+            component_id,
+            priority,
+        });
+
+        let mut replicaton_fns = self.world.resource_mut::<ReplicationFns>();
+        replicaton_fns.register_marker(marker_id);
+
+        self
+    }
+
+    unsafe fn register_marker_fns<M: Component, C: Component>(
+        &mut self,
+        write: WriteFn,
+        remove: RemoveFn,
+    ) -> &mut Self {
+        let component_id = self.world.init_component::<M>();
+        let command_markers = self.world.resource::<CommandMarkers>();
+        let marker_id = command_markers.marker_id(component_id);
+        self.world
+            .resource_scope(|world, mut replication_fns: Mut<ReplicationFns>| unsafe {
+                // SAFETY: The caller ensured that `write` can be safely called with a `SerdeFns` created for `C`.
+                replication_fns.register_marker_fns::<C>(world, marker_id, write, remove);
+            });
+
+        self
+    }
+}
+
+/// Registered markers that override functions if present for
+/// [`CommandFns`](super::replication_fns::command_fns::CommandFns).
+#[derive(Resource, Default)]
+pub(crate) struct CommandMarkers(Vec<CommandMarker>);
+
+impl CommandMarkers {
+    /// Inserts a new marker, maintaining sorting by their priority in descending order.
+    ///
+    /// Use [`ReplicationFns::register_marker`] to register a slot for command functions for this marker.
+    pub(super) fn insert(&mut self, marker: CommandMarker) -> CommandMarkerId {
+        let index = self
+            .0
+            .binary_search_by_key(&Reverse(marker.priority), |marker| Reverse(marker.priority))
+            .unwrap_or_else(|index| index);
+
+        self.0.insert(index, marker);
+
+        CommandMarkerId(index)
+    }
```

Comment:
We do invalidate old `CommandMarkerId`s, but we don't care about it since we are not storing them anywhere.
By calling `add_marker_slot` we don't repair IDs, we just insert a slot for a marker with `None` in the correct place based on priority. And the same thing when you set a function.

Should I write a comment about it in `CommandMarkerId` or maybe replace it with just `usize`?